lib/stages/BeamAssemble.cpp

```diff
@@ -0,0 +1,173 @@
+#include "StageFactory.hpp"   // for REGISTER_KOTEKAN_STAGE, StageMakerTemplate
+#include "Telescope.hpp"      // for Telescope
+#include "buffer.h"           // for get_metadata, mark_frame_empty, register_consumer, wait_fo...
+#include "kotekanLogging.hpp" // for INFO
+#include "visUtil.hpp"        // for frameID, modulo, ts_to_double
+#include "FrequencyAssembledMetadata.hpp"  // for FrequencyBin
+
+#include "fmt.hpp" // for format
+
+#include <map>      // for std::map
+#include <stdint.h> // for uint8_t, uint32_t
+#include <utility>  // for std::make_pair
+#include <functional> // for std::pair
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::Stage;
+
+REGISTER_KOTEKAN_STAGE(BeamAssemble);
+
+STAGE_CONSTRUCTOR(BeamAssemble) {
+    // retrieve th input buffer and register this class as a consumer for it
+    in_buf = get_buffer("in_buf");
+    register_consumer(in_buf, unique_name.c_str());
+    // retrieve the output buffer and register this class as a producer for it
+    out_buf = get_buffer("out_buf");
+    register_producer(out_buf, unique_name.c_str());
+    // retrieve the expected number of frequencies in the output
+    num_freq_per_output_frame = config.get<uint32_t>(unique_name, "num_freqs");
+    // retrieve the timeout in seconds for the arriving data in input buffer frames
+    arriving_data_timeout = config.get<uint32_t>(unique_name, "time_out");
+}
+
+BeamAssemble::~BeamAssemble() {}
+
+void BeamAssemble::main_thread() {
+
+    frameID in_frame_id(in_buf), out_frame_id(out_buf);
+    uint8_t *in_frame, *out_frame;
+
+    // the frame map used for timeout controlled frames
+    std::map<uint64_t, std::pair<uint8_t*, int>> out_frame_map;
+
+    // time stamp used for the frame map control
+    uint64_t current_frame_time_stamp;
+
+    // allocate new metadata objects for all the output buffer frames
+    for (int i = 0; i < out_buf->num_frames; i++) {
+        allocate_new_metadata_object(out_buf, i);
+    }
+
+    // thread infinite loop
+    while (!stop_thread) {
+
+        // acquire a frame from the input buffer
+        in_frame = wait_for_full_frame(in_buf, unique_name.c_str(), in_frame_id);
+        if (in_frame == nullptr)
+            break;
+
+        // retrieve the metadata and create a new metadata based on it
+        BeamMetadata* metadata = (BeamMetadata*)get_metadata(in_buf, in_frame_id);
+
+        // get the current time stamp in milliseconds from the current input frame metadata
+        current_frame_time_stamp = timespec_to_milliseconds(metadata->gps_time);
+
+        // loop over the open output frames and release the output frames if the timeout has reached
+        for (auto it = out_frame_map.begin(); it != out_frame_map.end();) {
+            // check if the time has passed more than arriving_data_timeout since this output frame
+            // has been opened
+            if (current_frame_time_stamp - it->first > arriving_data_timeout) {
```

Comment:
I don't think this should use the timestamp from the current input frame.  That value could be older than what's in the map, e.g. a delayed input frame received out of order.  In that case, then `current_frame_time_stamp - it->first` could wrap around since it's an unsigned value, and the newer frame would get expelled.

Reply:
Well, it exactly reflects the need for timeout tweak. Also, there are several approaches to do this: based on the out-of-window, mean of the gps_time, etc.

Reply:
I think the window should be based on the highest time value currently in the map, not the value you just received.  That value could be an old value coming from a delayed node, in that case you would get unexpected behavior, since the frames newer than that would get removed, because it's subtracting a smaller unsigned number by a larger one - resulting in a very large unsigned number. 

I think using this model without sorting in the output frame, the algorithm should look something like:
- Check if the new timestamp is older than the largest value in the map minus the timeout (the other comment).  Drop that frame if it's too old. 
- Add the new value to map if it doesn't exist in the map.
- If there was a change to the map, check if the map has any values smaller older that the highest time value minus the timeout and remove those entries.

Reply:
I had implemented exactly what you mentioned in my last commit. I fixed something there, and resubmitted. The first value of the map has the largest timestamp.

Again, the map is sorted, meaning that the frame with the lowest timestamp will be marked as full first (to be used in the next stage) and so on. Therefore, the frames will be release in ascending order in time and the next stage is responsible to process them in the same order to preserve it, if any ordering required there.

Reply:
The buffers act as rings and are processed in frame ID order, not the order in which they are released.    Releasing the frames in this way would sort if the buffer was a FIFO queue, but not with the current design.  Having said that, I think we can leave the reordering to the transpose/conversion from raw files to HDF5. 

However, I still think there is a problem on this line.  There is no guarantee that current_frame_time_stamp will be the latest timestamp.    If `current_frame_time_stamp < it->first` (e.g. if the timestamp value we just got is less than any value currently in the map), then the LHS of the inequality in the `if (current_frame_time_stamp - it->first > arriving_data_timeout)` statement will become a very large number (since it is unsigned), that will be greater than the timeout, and then inadvertently remove newer entries in the map.

Reply:
Well, adding another feature to the ring buffer is easier and more rational instead of carrying this need along to every stage or everywhere.

That's a very good point; and the solution to simply accompany another condition `( current_frame_timestamp > it->first)` with boolean `&&` operator or removing the unsigned from those types. I fixed and submitted.

Reply:
We can talk about it more in the meeting today, but changing the ring buffer is something we have to be very careful about doing.  We currently have many telescopes using it with dozens of stages.  Making changes there  requires substantial testing at this point.  It would also be sunk cost, since my current thinking is that we will add a new C++ version of the buffer which supports different operating modes (rings, FIFO, priority queues, and more), and then slowly replace the use of the older version in the existing systems, see #943  That is a project I want to do, but it requires a careful design proposal, a lot of modification to the core framework components, and likely several months of work.  From a project priority point of view, we should not start building a new buffer until we have at least completed the gated beams project and the core Outriggers functionality.